Mortar-based gap heat transfer

recuero opened the pull request:
- [X] Radiation
- [X] Conduction
- [X] 2D 
3D (Testing for next PR)

Refs. #19286

recuero committed:
Make small design changes to modular gap conductance - user object system

GapFluxModelBase is a friend of the modular gap conductance constraint, which
allows the user object base to cache geometry information from the constraint
that can be used by derived user objects implementing particular gap physics.

Refs. #19286

dschwen commented on modules/heat_conduction/include/constraints/ModularGapConductanceConstraint.h:
```diff
@@ -10,6 +10,7 @@
 #pragma once
 
 #include "ADMortarConstraint.h"
+#include "ModularGapConductanceConstraint.h"
```
Circular

dschwen replied:
You just need a forward, I think

```suggestion
class GapFluxModelBase;
```

dschwen commented on modules/heat_conduction/include/userobjects/GapFluxModelBase.h:
```diff
@@ -21,8 +22,20 @@ class GapFluxModelBase : public InterfaceUserObjectBase
   virtual void threadJoin(const UserObject &) final{};
+
+protected:
+  mutable unsigned int _qp;
```
Oh, I did not think of that....

This might not work reliably.

recuero committed:
Reinit displaced neighbor for IUO base (#19286)

recuero committed:
Add geometric computations for mortar gap heat transfer (#19286)

Also add a user object that mimics radiation as implemented in the
node on segment implementation. The idea is this user object can be
extended with more physics to provide a heat flux.

All this is to be tested.

recuero commented on framework/include/constraints/MortarConstraintBase.h:
```diff
@@ -108,11 +108,10 @@ class MortarConstraintBase : public Constraint,
                           const std::unordered_set<const Elem *> & inactive_lm_elems);
 
-private:
+protected:
   /// Reference to the finite element problem
   FEProblemBase & _fe_problem;
 
-protected:
```
@lindsayad acceptable?

lindsayad replied:
Do you need the non-const reference? If not, you could port what I've done here: https://github.com/idaholab/moose/pull/19266/files#diff-7345a9b3d222670aa12c4f58f967283bd34131dad796dbea8237031038ce84b4R111

recuero replied:
`const` ref is fine, that works.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/873723/) on 755d17d wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/19287/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/873750/) on 755d17d wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>4f272b</th>
      <th colspan=4>#19287 755d17</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>81.88%</td>
      <td>81.88%</td>
      <td>+0.00%</td>
      <td>50.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>71926</td>
      <td>71927</td>
      <td>+1</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15917</td>
      <td>15917</td>
      <td>-</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/19287/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/19287/coverage/framework)


## Modules coverage

### Heat conduction

<table>
  <thead>
    <tr>
      <th></th>
      <th>4f272b</th>
      <th colspan=4>#19287 755d17</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>87.94%</td>
      <td>85.77%</td>
      <td>-2.17%</td>
      <td>40.41%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>2743</td>
      <td>2791</td>
      <td>+48</td>
      <td>59</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>376</td>
      <td>463</td>
      <td>+87</td>
      <td>87</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/19287/coverage_diff/heat_conduction.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/19287/coverage/heat_conduction)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/19287/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/19287/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/19287/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/19287/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/19287/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/19287/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/19287/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/19287/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/19287/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/19287/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/19287/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/19287/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/19287/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/19287/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/19287/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/19287/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/19287/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/19287/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/19287/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/19287/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/19287/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/19287/coverage/tensor_mechanics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/19287/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 50.00% is less than the suggested 90.0%
- `heat_conduction` new line coverage rate 40.41% is less than the suggested 90.0%



This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/873750/) on 755d17d wanted to post the following:


The following coverage requirement(s) failed:
- `heat_conduction` coverage rate 85.77% is less than the required 87.0%

recuero committed:
Allow const access to _fe_problem from derived classes (#19286)

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/873839/) on 9c4ffb2 wanted to post the following:


The following coverage requirement(s) failed:
- `heat_conduction` coverage rate 85.77% is less than the required 87.0%

recuero committed:
Modify modular mortar heat transfer classes to capture heat transfer

Refs. #19286

recuero committed:
Add heat conduction UO (#19286)

recuero committed:
Add regression tests for mortar-based heat transfer (#19286)

Tests compare node-to-segment results for plate, cylinder, cylinder
axissymmetric and sphere problems. They include radiation and
conduction.

recuero committed:
Add documentation for user object conduction. (#19286)

Disallow renumbering of tests to check against csv temperature files.

recuero committed:
Add more tests (#19286)

 - Cover moose error lines of code
 - Set up problem with multiple heat flux Lagrange's multipliers

recuero committed:
Add test for gap cconductance function

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/878698/) on c13daf5 wanted to post the following:


The following coverage requirement(s) failed:
- `heat_conduction` coverage rate 86.29% is less than the required 87.0%

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/878698/) on c13daf5 wanted to post the following:


The following coverage requirement(s) failed:
- `heat_conduction` coverage rate 86.29% is less than the required 87.0%

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/878698/) on c13daf5 wanted to post the following:


The following coverage requirement(s) failed:
- `heat_conduction` coverage rate 86.86% is less than the required 87.0%

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/879618/) on d992b61 wanted to post the following:


The following coverage requirement(s) failed:
- `heat_conduction` coverage rate 86.86% is less than the required 87.0%

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/879618/) on d992b61 wanted to post the following:


The following coverage requirement(s) failed:
- `heat_conduction` coverage rate 86.86% is less than the required 87.0%

dschwen commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -3517,6 +3517,7 @@ FEProblemBase::addUserObject(const std::string & user_object_name,
     auto suo = std::dynamic_pointer_cast<SideUserObject>(user_object);
     auto isuo = std::dynamic_pointer_cast<InternalSideUserObject>(user_object);
     auto iuo = std::dynamic_pointer_cast<InterfaceUserObject>(user_object);
+    auto iuob = std::dynamic_pointer_cast<InterfaceUserObjectBase>(user_object);
```
```suggestion
    auto iuob = std::dynamic_pointer_cast<InterfaceUserObjectBase>(user_object);
```

dschwen replied:
if you can cast to `InterfaceUserObject` it's guaranteed that you can cast to `InterfaceUserObjectBase` so that's redundant

dschwen commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -3529,7 +3530,7 @@ FEProblemBase::addUserObject(const std::string & user_object_name,
         // shouldn't we add isuo
         _reinit_displaced_face = true;
-      else if (iuo)
+      else if (iuo || iuob)
```
```suggestion
      else if (iuob)
```

dschwen commented on modules/heat_conduction/src/constraints/ModularGapConductanceConstraint.C:
```diff
@@ -72,6 +96,203 @@ ModularGapConductanceConstraint::ModularGapConductanceConstraint(const InputPara
+  if (gap_geometry_type == GapGeometry::PLATE)
+  {
+    if (coord_sys == Moose::COORD_RSPHERICAL)
+      mooseError("'gap_geometry_type = PLATE' cannot be used with models having a spherical "
```
paramError?

recuero replied:
Yeah, all that is user's responsibility

dschwen commented on modules/heat_conduction/src/constraints/ModularGapConductanceConstraint.C:
```diff
@@ -72,6 +96,203 @@ ModularGapConductanceConstraint::ModularGapConductanceConstraint(const InputPara
+void
+ModularGapConductanceConstraint::computeGapRadii(const ADReal & gap_length)
+{
+  const Point current_point(_q_point[_qp]);
```
```suggestion
  const Point & current_point = _q_point[_qp];
```

dschwen commented on modules/heat_conduction/src/constraints/ModularGapConductanceConstraint.C:
```diff
@@ -97,14 +318,23 @@ ModularGapConductanceConstraint::computeQpResidual(Moose::MortarType mortar_type
+      _gap_width = (ad_phys_points_primary - ad_phys_points_secondary) * _normals[_qp];
+
+      // First, compute radii with _gap_width
+      computeGapRadii(_gap_width);
```
why pass it in when it's already a member?

dschwen commented on modules/heat_conduction/src/constraints/ModularGapConductanceConstraint.C:
```diff
@@ -97,14 +318,23 @@ ModularGapConductanceConstraint::computeQpResidual(Moose::MortarType mortar_type
+      computeGapRadii(_gap_width);
+
+      // With radii, compute adjusted gap length for conduction
+      _adjusted_length = computeGapLength();
```
same here, the function could just set them member

recuero replied:
Same idea, we see what we are doing here without diving into the member function. Then the member function can be `const`, making sure we are not messing other stuff.

dschwen replied:
ok

dschwen commented on modules/heat_conduction/src/userobjects/GapFluxModelConduction.C:
```diff
@@ -0,0 +1,96 @@
+  params.addRangeCheckedParam<Real>(
+      "min_gap", 1e-6, "min_gap>0", "A minimum gap (denominator) size");
+  params.addRangeCheckedParam<unsigned int>(
+      "min_gap_order",
```
should the min gap stuff be handled in the constraint maybe? Or is this really only relevant for the conduction flux model?

recuero replied:
I debated this... It's only used for conduction now, so it can be moved to the constraint if it ever becomes more general. Some of the computations in the constraint are only used for conduction too, but they are related to general problem parameters, so Ieft those in the constraint. The `min_gap` appears to be more conduction model-related.

dschwen reviewed (commented):
I think this looks really great. A lot clearer and easier to follow than the old implementation.

recuero committed:
Address Daniel's comments (#19286)
